File: mcp_servers/src/core/base-server.ts
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { Tool, Resource } from '@modelcontextprotocol/sdk/types.js';
import { createServerLogger } from './logger';
import { Logger } from 'pino';
import { MCPServerConfig } from './config';
import { EventEmitter } from 'events';

// Common interface definitions for cross-language compatibility
export interface MCPServerOptions {
  name: string;
  version: string;
  description: string;
  config?: Partial<MCPServerConfig>;
  capabilities?: ServerCapabilities;
}

export interface ServerCapabilities {
  tools?: boolean;
  resources?: boolean;
  prompts?: boolean;
  roots?: boolean;
  sampling?: boolean;
}

export interface ServerMetrics {
  uptime: number;
  requestCount: number;
  errorCount: number;
  toolCalls: number;
  resourceAccess: number;
  lastActivity: Date;
}

export interface HealthStatus {
  status: 'healthy' | 'degraded' | 'unhealthy';
  timestamp: Date;
  checks: HealthCheck[];
  metrics: ServerMetrics;
}

export interface HealthCheck {
  name: string;
  status: 'pass' | 'fail' | 'warn';
  message?: string;
  duration?: number;
}

export abstract class BaseMCPServer extends EventEmitter {
  protected server: Server;
  protected logger: Logger;
  protected tools: Map<string, Tool> = new Map();
  protected resources: Map<string, Resource> = new Map();
  protected metrics: ServerMetrics;
  protected healthChecks: Map<string, () => Promise<HealthCheck>> = new Map();
  protected isStarted: boolean = false;
  
  constructor(protected options: MCPServerOptions) {
    super();
    this.logger = createServerLogger(options.name);
    
    // Initialize metrics
    this.metrics = {
      uptime: 0,
      requestCount: 0,
      errorCount: 0,
      toolCalls: 0,
      resourceAccess: 0,
      lastActivity: new Date()
    };
    
    // Create server with enhanced capabilities
    this.server = new Server(
      {
        name: options.name,
        version: options.version,
      },
      {
        capabilities: {
          tools: options.capabilities?.tools ? {} : undefined,
          resources: options.capabilities?.resources ? {} : undefined,
          prompts: options.capabilities?.prompts ? {} : undefined,
          roots: options.capabilities?.roots ? {} : undefined,
          sampling: options.capabilities?.sampling ? {} : undefined,
        },
      }
    );
    
    this.setupErrorHandling();
    this.setupMetrics();
    this.registerStandardHealthChecks();
  }
  
  private setupErrorHandling(): void {
    this.server.onerror = (error) => {
      this.metrics.errorCount++;
      this.logger.error({ error }, 'Server error occurred');
      this.emit('error', error);
    };
    
    process.on('unhandledRejection', (reason, promise) => {
      this.metrics.errorCount++;
      this.logger.error({ reason, promise }, 'Unhandled rejection');
      this.emit('error', new Error(`Unhandled rejection: ${reason}`));
    });
    
    process.on('uncaughtException', (error) => {
      this.metrics.errorCount++;
      this.logger.error({ error }, 'Uncaught exception');
      this.emit('error', error);
      // Graceful shutdown instead of immediate exit
      this.gracefulShutdown();
    });
  }
  
  private setupMetrics(): void {
    const startTime = Date.now();
    
    // Update uptime every second
    setInterval(() => {
      this.metrics.uptime = Math.floor((Date.now() - startTime) / 1000);
    }, 1000);
  }
  
  private registerStandardHealthChecks(): void {
    // Memory usage check
    this.healthChecks.set('memory', async () => {
      const usage = process.memoryUsage();
      const heapUsedMB = Math.round(usage.heapUsed / 1024 / 1024);
      const heapTotalMB = Math.round(usage.heapTotal / 1024 / 1024);
      const usage_percent = (heapUsedMB / heapTotalMB) * 100;
      
      return {
        name: 'memory',
        status: usage_percent > 90 ? 'fail' : usage_percent > 75 ? 'warn' : 'pass',
        message: `Heap usage: ${heapUsedMB}MB/${heapTotalMB}MB (${usage_percent.toFixed(1)}%)`,
      };
    });
    
    // Server status check
    this.healthChecks.set('server', async () => {
      return {
        name: 'server',
        status: this.isStarted ? 'pass' : 'fail',
        message: this.isStarted ? 'Server is running' : 'Server is not started',
      };
    });
  }
  
  protected registerTool(tool: Tool): void {
    this.tools.set(tool.name, tool);
    this.logger.info({ toolName: tool.name }, 'Tool registered');
    this.emit('tool_registered', tool);
  }
  
  protected registerResource(resource: Resource): void {
    this.resources.set(resource.uri, resource);
    this.logger.info({ resourceUri: resource.uri }, 'Resource registered');
    this.emit('resource_registered', resource);
  }
  
  protected addHealthCheck(name: string, check: () => Promise<HealthCheck>): void {
    this.healthChecks.set(name, check);
    this.logger.info({ checkName: name }, 'Health check registered');
  }
  
  public async start(): Promise<void> {
    try {
      // Setup server-specific components
      await this.setupTools();
      await this.setupResources();
      
      this.setupRequestHandlers();
      
      // Start the server
      const transport = new StdioServerTransport();
      await this.server.connect(transport);
      
      this.isStarted = true;
      this.emit('started');
      
      this.logger.info(
        { 
          name: this.options.name, 
          version: this.options.version,
          toolCount: this.tools.size,
          resourceCount: this.resources.size
        }, 
        'MCP server started'
      );
    } catch (error) {
      this.logger.error({ error }, 'Failed to start server');
      this.emit('start_failed', error);
      throw error;
    }
  }
  
  private setupRequestHandlers(): void {
    // Tools handlers
    this.server.setRequestHandler('tools/list' as any, async () => {
      this.metrics.requestCount++;
      this.metrics.lastActivity = new Date();
      return { tools: Array.from(this.tools.values()) };
    });
    
    this.server.setRequestHandler('tools/call' as any, async (request: any) => {
      this.metrics.requestCount++;
      this.metrics.toolCalls++;
      this.metrics.lastActivity = new Date();
      
      const { name, arguments: args } = request.params as {
        name: string;
        arguments?: unknown;
      };
      
      const tool = this.tools.get(name);
      if (!tool) {
        this.metrics.errorCount++;
        throw new Error(`Tool ${name} not found`);
      }
      
      this.logger.info({ toolName: name, args }, 'Tool called');
      
      try {
        const result = await this.executeTool(name, args);
        this.emit('tool_executed', { name, args, result });
        return { content: [{ type: 'text', text: JSON.stringify(result) }] };
      } catch (error) {
        this.metrics.errorCount++;
        this.logger.error({ error, toolName: name }, 'Tool execution failed');
        this.emit('tool_failed', { name, args, error });
        throw error;
      }
    });
    
    // Resources handlers
    this.server.setRequestHandler('resources/list' as any, async () => {
      this.metrics.requestCount++;
      this.metrics.lastActivity = new Date();
      return { resources: Array.from(this.resources.values()) };
    });
    
    this.server.setRequestHandler('resources/read' as any, async (request: any) => {
      this.metrics.requestCount++;
      this.metrics.resourceAccess++;
      this.metrics.lastActivity = new Date();
      
      const { uri } = request.params as { uri: string };
      
      const resource = this.resources.get(uri);
      if (!resource) {
        this.metrics.errorCount++;
        throw new Error(`Resource ${uri} not found`);
      }
      
      try {
        const content = await this.readResource(uri);
        this.emit('resource_read', { uri, content });
        return { contents: [content] };
      } catch (error) {
        this.metrics.errorCount++;
        this.logger.error({ error, uri }, 'Resource read failed');
        this.emit('resource_failed', { uri, error });
        throw error;
      }
    });
  }
  
  public async getHealth(): Promise<HealthStatus> {
    const checks: HealthCheck[] = [];
    
    for (const [name, checkFn] of this.healthChecks) {
      try {
        const start = Date.now();
        const check = await checkFn();
        check.duration = Date.now() - start;
        checks.push(check);
      } catch (error) {
        checks.push({
          name,
          status: 'fail',
          message: `Health check failed: ${error instanceof Error ? error.message : String(error)}`,
        });
      }
    }
    
    const hasFailures = checks.some(check => check.status === 'fail');
    const hasWarnings = checks.some(check => check.status === 'warn');
    
    return {
      status: hasFailures ? 'unhealthy' : hasWarnings ? 'degraded' : 'healthy',
      timestamp: new Date(),
      checks,
      metrics: { ...this.metrics },
    };
  }
  
  public async gracefulShutdown(): Promise<void> {
    this.logger.info('Starting graceful shutdown');
    this.emit('shutdown_started');
    
    try {
      // Perform cleanup operations
      await this.cleanup();
      
      // Close server connections
      if (this.server) {
        await this.server.close();
      }
      
      this.isStarted = false;
      this.emit('shutdown_complete');
      this.logger.info('Graceful shutdown completed');
    } catch (error) {
      this.logger.error({ error }, 'Error during graceful shutdown');
      this.emit('shutdown_error', error);
    } finally {
      process.exit(0);
    }
  }
  
  // Abstract methods that must be implemented by concrete servers
  protected abstract setupTools(): Promise<void>;
  protected abstract setupResources(): Promise<void>;
  protected abstract executeTool(name: string, args: unknown): Promise<unknown>;
  protected abstract readResource(uri: string): Promise<{ uri: string; mimeType?: string; text?: string; blob?: string }>;
  protected abstract cleanup(): Promise<void>;
}

File: mcp_servers/src/core/base_server.py
"""
MCP Server Base Class - Python Implementation

Standard base class for all Python MCP servers providing consistent
architecture patterns, error handling, logging, and health monitoring.
"""

import asyncio
import logging
import time
import traceback
from abc import ABC, abstractmethod
from datetime import datetime
from typing import Any, Dict, List, Optional, Callable, Union
from dataclasses import dataclass, field
from enum import Enum
import json
import psutil
import signal

from mcp import Tool, Resource
from mcp.server import Server
from mcp.server.stdio import stdio_server


class HealthStatus(Enum):
    """Health status enumeration"""
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class CheckStatus(Enum):
    """Health check status enumeration"""
    PASS = "pass"
    FAIL = "fail"
    WARN = "warn"


@dataclass
class ServerCapabilities:
    """Server capabilities configuration"""
    tools: bool = True
    resources: bool = True
    prompts: bool = False
    roots: bool = False
    sampling: bool = False


@dataclass
class ServerMetrics:
    """Server performance metrics"""
    uptime: int = 0
    request_count: int = 0
    error_count: int = 0
    tool_calls: int = 0
    resource_access: int = 0
    last_activity: datetime = field(default_factory=datetime.now)


@dataclass
class HealthCheck:
    """Individual health check result"""
    name: str
    status: CheckStatus
    message: Optional[str] = None
    duration: Optional[float] = None


@dataclass
class HealthReport:
    """Complete health status report"""
    status: HealthStatus
    timestamp: datetime
    checks: List[HealthCheck]
    metrics: ServerMetrics


@dataclass
class MCPServerOptions:
    """MCP Server configuration options"""
    name: str
    version: str
    description: str
    config: Optional[Dict[str, Any]] = None
    capabilities: Optional[ServerCapabilities] = None


class BaseMCPServer(ABC):
    """
    Base class for all MCP servers providing standardized:
    - Error handling and logging
    - Health monitoring and metrics
    - Request/response patterns
    - Graceful shutdown
    - Event emission
    """
    
    def __init__(self, options: MCPServerOptions):
        self.options = options
        self.logger = self._setup_logging()
        self.server = Server(options.name)
        
        # Initialize state
        self.tools: Dict[str, Tool] = {}
        self.resources: Dict[str, Resource] = {}
        self.metrics = ServerMetrics()
        self.health_checks: Dict[str, Callable[[], HealthCheck]] = {}
        self.is_started = False
        self.start_time = time.time()
        
        # Event callbacks
        self.event_callbacks: Dict[str, List[Callable]] = {}
        
        # Setup standard components
        self._setup_error_handling()
        self._setup_metrics()
        self._register_standard_health_checks()
        self._setup_signal_handlers()
    
    def _setup_logging(self) -> logging.Logger:
        """Setup structured logging"""
        logger = logging.getLogger(self.options.name)
        logger.setLevel(logging.INFO)
        
        if not logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
            )
            handler.setFormatter(formatter)
            logger.addHandler(handler)
        
        return logger
    
    def _setup_error_handling(self) -> None:
        """Setup comprehensive error handling"""
        def handle_exception(exc_type, exc_value, exc_traceback):
            if issubclass(exc_type, KeyboardInterrupt):
                return
            
            self.metrics.error_count += 1
            self.logger.error(
                "Uncaught exception",
                exc_info=(exc_type, exc_value, exc_traceback)
            )
            self.emit('error', exc_value)
    
    def _setup_metrics(self) -> None:
        """Setup metrics collection"""
        async def update_metrics():
            while self.is_started:
                self.metrics.uptime = int(time.time() - self.start_time)
                await asyncio.sleep(1)
        
        # Start metrics update task when server starts
        self._metrics_task = None
    
    def _register_standard_health_checks(self) -> None:
        """Register standard health checks"""
        
        def memory_check() -> HealthCheck:
            """Check memory usage"""
            try:
                process = psutil.Process()
                memory_info = process.memory_info()
                memory_percent = process.memory_percent()
                
                status = CheckStatus.PASS
                if memory_percent > 90:
                    status = CheckStatus.FAIL
                elif memory_percent > 75:
                    status = CheckStatus.WARN
                
                return HealthCheck(
                    name="memory",
                    status=status,
                    message=f"Memory usage: {memory_info.rss / 1024 / 1024:.1f}MB ({memory_percent:.1f}%)"
                )
            except Exception as e:
                return HealthCheck(
                    name="memory",
                    status=CheckStatus.FAIL,
                    message=f"Memory check failed: {str(e)}"
                )
        
        def server_check() -> HealthCheck:
            """Check server status"""
            return HealthCheck(
                name="server",
                status=CheckStatus.PASS if self.is_started else CheckStatus.FAIL,
                message="Server is running" if self.is_started else "Server is not started"
            )
        
        self.health_checks["memory"] = memory_check
        self.health_checks["server"] = server_check
    
    def _setup_signal_handlers(self) -> None:
        """Setup signal handlers for graceful shutdown"""
        def signal_handler(signum, frame):
            self.logger.info(f"Received signal {signum}, initiating graceful shutdown")
            asyncio.create_task(self.graceful_shutdown())
        
        signal.signal(signal.SIGTERM, signal_handler)
        signal.signal(signal.SIGINT, signal_handler)
    
    def register_tool(self, tool: Tool) -> None:
        """Register a tool with the server"""
        self.tools[tool.name] = tool
        self.logger.info(f"Tool registered: {tool.name}")
        self.emit('tool_registered', tool)
    
    def register_resource(self, resource: Resource) -> None:
        """Register a resource with the server"""
        self.resources[resource.uri] = resource
        self.logger.info(f"Resource registered: {resource.uri}")
        self.emit('resource_registered', resource)
    
    def add_health_check(self, name: str, check: Callable[[], HealthCheck]) -> None:
        """Add a custom health check"""
        self.health_checks[name] = check
        self.logger.info(f"Health check registered: {name}")
    
    def on(self, event: str, callback: Callable) -> None:
        """Register event callback"""
        if event not in self.event_callbacks:
            self.event_callbacks[event] = []
        self.event_callbacks[event].append(callback)
    
    def emit(self, event: str, *args, **kwargs) -> None:
        """Emit event to registered callbacks"""
        if event in self.event_callbacks:
            for callback in self.event_callbacks[event]:
                try:
                    callback(*args, **kwargs)
                except Exception as e:
                    self.logger.error(f"Error in event callback for {event}: {str(e)}")
    
    async def start(self) -> None:
        """Start the MCP server"""
        try:
            # Setup server-specific components
            await self.setup_tools()
            await self.setup_resources()
            
            self._setup_request_handlers()
            
            # Start metrics collection
            self._metrics_task = asyncio.create_task(self._update_metrics_loop())
            
            # Start the server
            async with stdio_server() as streams:
                await self.server.run(streams[0], streams[1])
            
            self.is_started = True
            self.emit('started')
            
            self.logger.info(
                f"MCP server started: {self.options.name} v{self.options.version} "
                f"(tools: {len(self.tools)}, resources: {len(self.resources)})"
            )
        
        except Exception as error:
            self.logger.error(f"Failed to start server: {str(error)}")
            self.emit('start_failed', error)
            raise
    
    def _setup_request_handlers(self) -> None:
        """Setup standard request handlers"""
        
        @self.server.list_tools()
        async def list_tools():
            """List available tools"""
            self.metrics.request_count += 1
            self.metrics.last_activity = datetime.now()
            return list(self.tools.values())
        
        @self.server.call_tool()
        async def call_tool(name: str, arguments: Optional[Dict[str, Any]] = None):
            """Execute a tool"""
            self.metrics.request_count += 1
            self.metrics.tool_calls += 1
            self.metrics.last_activity = datetime.now()
            
            if name not in self.tools:
                self.metrics.error_count += 1
                raise ValueError(f"Tool {name} not found")
            
            self.logger.info(f"Tool called: {name}")
            
            try:
                result = await self.execute_tool(name, arguments or {})
                self.emit('tool_executed', name, arguments, result)
                return [{"type": "text", "text": json.dumps(result)}]
            except Exception as error:
                self.metrics.error_count += 1
                self.logger.error(f"Tool execution failed: {name} - {str(error)}")
                self.emit('tool_failed', name, arguments, error)
                raise
        
        @self.server.list_resources()
        async def list_resources():
            """List available resources"""
            self.metrics.request_count += 1
            self.metrics.last_activity = datetime.now()
            return list(self.resources.values())
        
        @self.server.read_resource()
        async def read_resource(uri: str):
            """Read a resource"""
            self.metrics.request_count += 1
            self.metrics.resource_access += 1
            self.metrics.last_activity = datetime.now()
            
            if uri not in self.resources:
                self.metrics.error_count += 1
                raise ValueError(f"Resource {uri} not found")
            
            try:
                content = await self.read_resource_content(uri)
                self.emit('resource_read', uri, content)
                return [content]
            except Exception as error:
                self.metrics.error_count += 1
                self.logger.error(f"Resource read failed: {uri} - {str(error)}")
                self.emit('resource_failed', uri, error)
                raise
    
    async def _update_metrics_loop(self) -> None:
        """Continuous metrics update loop"""
        while self.is_started:
            self.metrics.uptime = int(time.time() - self.start_time)
            await asyncio.sleep(1)
    
    async def get_health(self) -> HealthReport:
        """Get comprehensive health status"""
        checks = []
        
        for name, check_fn in self.health_checks.items():
            try:
                start_time = time.time()
                check = check_fn()
                check.duration = time.time() - start_time
                checks.append(check)
            except Exception as e:
                checks.append(HealthCheck(
                    name=name,
                    status=CheckStatus.FAIL,
                    message=f"Health check failed: {str(e)}"
                ))
        
        # Determine overall status
        has_failures = any(check.status == CheckStatus.FAIL for check in checks)
        has_warnings = any(check.status == CheckStatus.WARN for check in checks)
        
        if has_failures:
            status = HealthStatus.UNHEALTHY
        elif has_warnings:
            status = HealthStatus.DEGRADED
        else:
            status = HealthStatus.HEALTHY
        
        return HealthReport(
            status=status,
            timestamp=datetime.now(),
            checks=checks,
            metrics=self.metrics
        )
    
    async def graceful_shutdown(self) -> None:
        """Perform graceful shutdown"""
        self.logger.info("Starting graceful shutdown")
        self.emit('shutdown_started')
        
        try:
            # Stop metrics collection
            if self._metrics_task:
                self._metrics_task.cancel()
                try:
                    await self._metrics_task
                except asyncio.CancelledError:
                    pass
            
            # Perform cleanup
            await self.cleanup()
            
            self.is_started = False
            self.emit('shutdown_complete')
            self.logger.info("Graceful shutdown completed")
        
        except Exception as error:
            self.logger.error(f"Error during graceful shutdown: {str(error)}")
            self.emit('shutdown_error', error)
    
    # Abstract methods that must be implemented by concrete servers
    @abstractmethod
    async def setup_tools(self) -> None:
        """Setup server-specific tools"""
        pass
    
    @abstractmethod
    async def setup_resources(self) -> None:
        """Setup server-specific resources"""
        pass
    
    @abstractmethod
    async def execute_tool(self, name: str, arguments: Dict[str, Any]) -> Any:
        """Execute a tool with given arguments"""
        pass
    
    @abstractmethod
    async def read_resource_content(self, uri: str) -> Dict[str, Any]:
        """Read resource content"""
        pass
    
    @abstractmethod
    async def cleanup(self) -> None:
        """Perform cleanup operations before shutdown"""
        pass